src/sinks/util/udp.rs

```diff
@@ -104,201 +117,166 @@ impl UdpConnector {
         let addr = SocketAddr::new(ip, self.port);
         let bind_address = find_bind_address(&addr);
 
-        let socket = UdpSocket::bind(bind_address).context(BindError)?;
-        socket.connect(addr).context(ConnectError)?;
+        let socket = UdpSocket::bind(bind_address).await.context(BindError)?;
+        socket.connect(addr).await.context(ConnectError)?;
 
         Ok(socket)
     }
 
-    async fn connect_backoff(self) -> UdpSocket {
-        let mut backoff = ExponentialBackoff::from_millis(2)
-            .factor(250)
-            .max_delay(Duration::from_secs(60));
-
+    async fn connect_backoff(&self) -> UdpSocket {
+        let mut backoff = Self::fresh_backoff();
         loop {
             match self.connect().await {
-                Ok(socket) => return socket,
+                Ok(socket) => {
+                    emit!(UdpSocketConnectionEstablished {});
+                    return socket;
+                }
                 Err(error) => {
-                    error!(message = "Unable to connect UDP socket.", %error);
+                    emit!(UdpSocketConnectionFailed { error });
                     delay_for(backoff.next().unwrap()).await;
                 }
             }
         }
     }
 
-    fn healthcheck(&self) -> BoxFuture<'static, crate::Result<()>> {
-        let this = self.clone();
-        async move { this.connect().await }
-            .map_ok(|_| ())
-            .err_into()
-            .boxed()
-    }
-}
-
-impl Into<UdpSink> for UdpConnector {
-    fn into(self) -> UdpSink {
-        UdpSink::new(self.host, self.port, self.resolver)
+    async fn healthcheck(&self) -> crate::Result<()> {
+        self.connect().await.map(|_| ()).map_err(Into::into)
     }
 }
 
 enum UdpServiceState {
+    Disconnected,
     Connecting(BoxFuture<'static, UdpSocket>),
     Connected(UdpSocket),
+    Sending(oneshot::Receiver<UdpSocket>),
 }
 
 pub struct UdpService {
-    state: UdpServiceState,
+    connector: UdpConnector,
+    state: Cell<UdpServiceState>,
```

Comment:
`Cell` is not needed

```diff
diff --git a/src/sinks/util/udp.rs b/src/sinks/util/udp.rs
index ed5217a1f..e4d3c2fc9 100644
--- a/src/sinks/util/udp.rs
+++ b/src/sinks/util/udp.rs
@@ -16,7 +16,6 @@ use futures::{future::BoxFuture, ready, stream::BoxStream, FutureExt, StreamExt}
 use serde::{Deserialize, Serialize};
 use snafu::{ResultExt, Snafu};
 use std::{
-    cell::Cell,
     net::{IpAddr, Ipv4Addr, Ipv6Addr, SocketAddr},
     pin::Pin,
     sync::Arc,
@@ -153,14 +152,14 @@ enum UdpServiceState {
 
 pub struct UdpService {
     connector: UdpConnector,
-    state: Cell<UdpServiceState>,
+    state: UdpServiceState,
 }
 
 impl UdpService {
     fn new(connector: UdpConnector) -> Self {
         Self {
             connector,
-            state: Cell::new(UdpServiceState::Disconnected),
+            state: UdpServiceState::Disconnected,
         }
     }
 }
@@ -172,7 +171,7 @@ impl tower::Service<Bytes> for UdpService {
 
     fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
         loop {
-            self.state = Cell::new(match self.state.get_mut() {
+            self.state = match &mut self.state {
                 UdpServiceState::Disconnected => {
                     let connector = self.connector.clone();
                     UdpServiceState::Connecting(Box::pin(async move {
@@ -191,7 +190,7 @@ impl tower::Service<Bytes> for UdpService {
                     };
                     UdpServiceState::Connected(socket)
                 }
-            });
+            };
         }
         Poll::Ready(Ok(()))
     }
@@ -199,10 +198,11 @@ impl tower::Service<Bytes> for UdpService {
     fn call(&mut self, msg: Bytes) -> Self::Future {
         let (sender, receiver) = oneshot::channel();
 
-        let mut socket = match self.state.replace(UdpServiceState::Sending(receiver)) {
-            UdpServiceState::Connected(socket) => socket,
-            _ => panic!("UdpService::poll_ready should be called first"),
-        };
+        let mut socket =
+            match std::mem::replace(&mut self.state, UdpServiceState::Sending(receiver)) {
+                UdpServiceState::Connected(socket) => socket,
+                _ => panic!("UdpService::poll_ready should be called first"),
+            };
 
         Box::pin(async move {
             // TODO: Add reconnect support as TCP/Unix?
```

Reply:
Wow, thanks! I did not use `std:mem` previously.